client/snapclient.1
.\"groff -Tascii -man snapclient.1
.TH SNAPCLIENT 1 "September 2025"
.SH NAME
snapclient - Snapcast client
.SH SYNOPSIS
\fBsnapclient\fR [\fIOPTION\fR]...
.SH DESCRIPTION
Snapcast is a multiroom client-server audio player, where all clients are
time synchronized with the server to play perfectly synced audio. It's not a
standalone player, but an extension that turns your existing audio player into
a Sonos-like multiroom solution.
.br
The server's audio input is a named pipe `/tmp/snapfifo`. All data that is fed
into this file will be send to the connected clients. One of the most generic
ways to use Snapcast is in conjunction with the music player daemon or Mopidy,
which can be configured to use a named pipe as audio output.
.SS Usage: snapclient [options...] [url]

 With 'url' = <tcp|ws|wss>://<snapserver host or IP or mDNS service name>[:port]
 For example: 'tcp://192.168.1.1:1704', or 'ws://)homeserver.local' or 'wss://_snapcast-https._tcp'
 If 'url' is not configured, snapclient defaults to 'tcp://_snapcast._tcp'
:
.TP
\fB--help\fR
Produce help message
.TP
\fB-v, --version\fR
Show version number
.TP
\fB-h, --host arg\fR
(deprecated, use [url]) Server hostname or ip address
.TP
\fB-p, --port arg (=1704)\fR
(deprecated, use [url]) Server port
.TP
\fB-i, --instance arg (=1)\fR
Instance id when running multiple instances on the same host
.TP
\fB--hostID arg\fR
Unique host id, default is MAC address
.TP
\fB--cert arg (="")\fR
Client certificate file (PEM format)
.TP
\fB--cert-key arg (="")\fR
Client private key file (PEM format)
.TP
\fB--key-password arg\fR
Key password (for encrypted private key)
.TP
\fB--server-cert [=arg(="default certificates")]\fR
Verify server with CA certificate (PEM format)
.TP
\fB-l, --list\fR
List PCM devices
.TP
\fB-s, --soundcard arg (=default)\fR
Index or name of the PCM device
.TP
\fB--latency arg (=0)\fR
Latency of the PCM device
.TP
\fB--sampleformat arg\fR
Resample audio stream to <rate>:<bits>:<channels>
.TP
\fB--player arg (=alsa)\fR
alsa|pipewire|file[:<options>|?]
.TP
\fB--mixer arg (=software)\fR
software|hardware|script|none|?[:<options>]
.TP
\fB-d, --daemon [=arg(=-3)]\fR
Daemonize, optional process priority [-20..19]
.TP
\fB--user arg\fR
The user[:group] to run snapclient as when daemonized
.TP
\fB--logsink arg\fR
Log sink [null,system,stdout,stderr,file:<filename>]
.TP
\fB--logfilter arg (=*:info)\fR
Log filter <tag>:<level>[,<tag>:<level>]* with tag = * or <log tag> and level = [trace,debug,info,notice,warning,error,fatal]
.SH FILES
.TP
\fI/etc/default/snapclient\fR
the daemon default configuration file
.SH "COPYRIGHT"
Copyright (C) 2014-2025 Johannes Pohl (snapcast@badaix.de).
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
snapserver(1)


changelog.md
# Snapcast changelog

## Version 0.33.0

### Features

- Server: Add PipeWire stream reader (PR #1418, Issue #1371)
- Client: Add PipeWire player (PR #1420, Issue #1419, #821)
- Server: process stream sources can be added via RPC (Issue #1359, PR #1444)
- Server: mDNS can be disabled (Issue #1172, PR #1437)
- Server: Audio streaming over plain TCP (port 1704) can be disabled (PR #1437)
- Server: Add control script for `go-librespot` (PR #1433, Issue #1424)

### Bugfixes

- Client: Fix case of command line arguments (Issue #1411)
- Client: Fix deadlock during shutdown (PR #1415)
- Fix building with boost v1.89.0 (Issues #1413)
- Fix MacOS CI (PR #1446)

### General

- Server: Log warning if default config is missing (PR #1402)
- Server: Make SSL support optional (`-DBUILD_WITH_SSL`) (Issue #1426, PR #1427)
- Make ALSA optional (PR #1422)
- Enable OpenSSL support for Android builds in `CMakeLists.txt` (PR #1373)
- Update Readme (PR #1431)
- Build Trixie packages for amd64 (Issue #1429)
- `tcp` section in `snapserver.conf` renamed to `tcp-control` (PR #1437)
- TCP streaming settings moved from `stream` to `tcp-streaming` (PR #1437)

### Contributors

- @badaix
- @mafrosis
- @aanno
- @filmor
- @sajjadsabzkar
- @MickMorley
- @lucianm

_Johannes Pohl <snapcast@badaix.de>  Tue, 23 Sep 2025 00:13:37 +0200_

## Version 0.32.3

### Features

### Bugfixes

- Server: re-add port to host in json RPC stream requests (Issue #1410)
- Documentation: Update Windows build instructions (PR #1408)

### General

### Contributors

- @stijnvdb88
- @badaix

_Johannes Pohl <snapcast@badaix.de>  Fri, 08 Aug 2025 00:13:37 +0200_

## Version 0.32.2

### Bugfixes

- Fix building Snapclient debian package with Pulse audio (Issue #1406)

### Contributors

- @badaix

_Johannes Pohl <snapcast@badaix.de>  Tue, 05 Aug 2025 00:13:37 +0200_

## Version 0.32.1

### Bugfixes

- Add OpenSSL dlls to Windows bundle (Issue #1405)

### Contributors

- @badaix

_Johannes Pohl <snapcast@badaix.de>  Mon, 04 Aug 2025 00:13:37 +0200_

## Version 0.32.0

### Features

- Client: Add support for (secure-) websockets (Issue #1325, PR #1340)
- Server: Add client certificate validation (Issue #1334)
- Server: Add control script for Plex (PR #1346)

### Bugfixes

- Fix typos (PR #1333, PR #1341, PR #1345)
- Add `mpv` option `--ao-pcm-waveheader=no` on player setup page (Issue #1094)
- Fix clang-tidy static analysis issues (Issue #1354)
- Fix building with Jack on Fedora (Issue #1363)
- Snapserver silently ignores if config file cannot be read (Issue #1402) 
- Add required packages to build instructions (Issue #1397)
- Fix stream URI string encoding (PR #1398)
- Fix reading port for TCP streams (PR #1399)

### General

- Snapweb: Update to v0.9.1
- Client: Command line arguments '--host' and '--port' are deprecated
- Update binary_protocol.md (Issue #1339)
- Add missing doxygen code documentation (Issue #1352)
- Add clang-tidy static analysis to CI (Issue #1353)
- Compiles with boost v1.88.0 (Issue #1367, PR #1378, PR #1387)
- Add Music Assistant project to readme (PR #1366)
 
### Contributors

- @chicco-carone
- @rnestler
- @badaix
- @chriscn
- @SantiagoSotoC
- @tobster
- @gnumpi

_Johannes Pohl <snapcast@badaix.de>  Sun, 03 Aug 2025 00:13:37 +0200_

## Version 0.31.0

### Features

- Server: Add secure RPC method `Stream.AddStream` (PR #1331, Issue #1330)
- Server: Re-add RPC method `RemoveStream` (Issue #1330)

### Bugfixes

- Fix crash when calling `Stream.RemoveStream`
- Fix typo in `changelog.md` (PR #1328)
- Fix typo in `configuration.md` (PR #1329)
- Update to AixLog v1.5.1, flushing log lines (PR #1332)

### General

- Add more doxygen documentation

_Johannes Pohl <snapcast@badaix.de>  Thu, 23 Jan 2025 00:13:37 +0200_

## Version 0.30.0

### Features

- Server: Add support for HTTPS and WSS (Issue #1322)
- Server: custom URL prefix for HTTP URLs (PR #1260)
- Server: listen to both IPv4 and IPv6 for incoming connections (PR #1290)

### Bugfixes

- Fix CVE-2023-36177: Remove "AddStream" and "RemoveStream" RPC (Issue #1327)
- Fix compile error on MacOS
- Remove PiMusicBox from readme (Issue ##1304)
- Add missing include (Issue #1295)
- Fix building with boost v1.87.0 (Issues #1305, #1306, #1307, PR #1308)
- Log tag is missing in syslog (Issue #1276)

### General

- CI: execute unit tests (asan and tsan)
- CI: create doxygen documentation
- CI: compile release and debug builds
- Code cleanup
- Add more doxygen documentation

_Johannes Pohl <snapcast@badaix.de>  Fri, 17 Jan 2025 00:13:37 +0200_

## Version 0.29.0

### Features

- Server: Add support for Jack streams (PR #1241, Issue #644)
- Client: Add support for 24 and 32 bit audio on Android

### Bugfixes

- Fix linker paths on ARM based MacOS

### General

- Snapweb: Update to v0.8.0
- CI: Add clang 16, 17, 18, gcc 14, xcode 15.1, 15.2, 15.3, 15.4
- CI: Create GitHub draft release

_Johannes Pohl <snapcast@badaix.de>  Sun, 25 Aug 2024 00:13:37 +0200_

## Version 0.28.0

### Features

- Server: Use filename as title for FileStreams
- Client: Loop through resolved server IPs during connect (Issue #1216)

### Bugfixes

- Server: Fix noise caused by reading half samples (Issue #1170)
- Server: Check open syscall error in PipeStream::do_connect (PR #1150)
- Server: Tweak Airplay support (#1102)
- Server: Lower log severity for shairport-sync (Issue #891)
- Server: Emits notifications for streams with codec=null (Is11sue #1205)
- Server: Fix delayed switching of meta streams (Issue #1187)
- Server: Mixer script calls piling up (Issue #1174)
- Server: Mixer script creates zombie processes (Issue #1157)
- Improve connect handling in meta_mopidy.py
- Improve connect handling in meta_librespot-java.py
- Fix cppcheck issues

### General

- CI: Build debian bookworm and bullseye packages for amd64, arm and arm64
- CI: Add cppcheck static analysis
- Update documentation (PR #1130, PR #1067)
- Delete deprecated Makefiles, CMake is the only supported buildsystem
- Snapweb: Update to v0.7.0


client/snapClient.cpp
/***
    This file is part of snapcast
    Copyright (C) 2014-2019  Johannes Pohl

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
***/

#include <iostream>
#include <sys/resource.h>

#include "browseZeroConf/browsemDNS.h"
#include "common/popl.hpp"
#include "controller.h"

#ifdef HAS_ALSA
#include "player/alsaPlayer.h"
#endif
#ifdef HAS_DAEMON
#include "common/daemon.h"
#endif
#include "common/aixlog.hpp"
#include "common/signalHandler.h"
#include "common/strCompat.h"
#include "common/utils.h"
#include "metadata.h"


using namespace std;
using namespace popl;

volatile sig_atomic_t g_terminated = false;

PcmDevice getPcmDevice(const std::string& soundcard)
{
#ifdef HAS_ALSA
    vector<PcmDevice> pcmDevices = AlsaPlayer::pcm_list();

    try
    {
        int soundcardIdx = cpt::stoi(soundcard);
        for (auto dev : pcmDevices)
            if (dev.idx == soundcardIdx)
                return dev;
    }
    catch (...)
    {
    }

    for (auto dev : pcmDevices)
        if (dev.name.find(soundcard) != string::npos)
            return dev;
#endif
    PcmDevice pcmDevice;
    return pcmDevice;
}


int main(int argc, char** argv)
{
#ifdef MACOS
#pragma message "Warning: the macOS support is experimental and might not be maintained"
#endif
    int exitcode = EXIT_SUCCESS;
    try
    {
        string meta_script("");
        string soundcard("default");
        string host("");
        size_t port(1704);
        int latency(0);
        size_t instance(1);

        OptionParser op("Allowed options");
        auto helpSwitch = op.add<Switch>("", "help", "produce help message");
        auto groffSwitch = op.add<Switch, Attribute::hidden>("", "groff", "produce groff message");
        auto debugOption = op.add<Implicit<string>, Attribute::hidden>("", "debug", "enable debug logging", "");
        auto versionSwitch = op.add<Switch>("v", "version", "show version number");
#if defined(HAS_ALSA)
        auto listSwitch = op.add<Switch>("l", "list", "list pcm devices");
        /*auto soundcardValue =*/op.add<Value<string>>("s", "soundcard", "index or name of the soundcard", "default", &soundcard);
#endif
        auto metaStderr = op.add<Switch>("e", "mstderr", "send metadata to stderr");
        // auto metaHook =       op.add<Value<string>>("m", "mhook", "script to call on meta tags", "", &meta_script);
        /*auto hostValue =*/op.add<Value<string>>("h", "host", "server hostname or ip address", "", &host);
        /*auto portValue =*/op.add<Value<size_t>>("p", "port", "server port", 1704, &port);
#ifdef HAS_DAEMON
        int processPriority(-3);
        auto daemonOption = op.add<Implicit<int>>("d", "daemon", "daemonize, optional process priority [-20..19]", -3, &processPriority);
        auto userValue = op.add<Value<string>>("", "user", "the user[:group] to run snapclient as when daemonized");
#endif
        /*auto latencyValue =*/op.add<Value<int>>("", "latency", "latency of the soundcard", 0, &latency);
        /*auto instanceValue =*/op.add<Value<size_t>>("i", "instance", "instance id", 1, &instance);
        auto hostIdValue = op.add<Value<string>>("", "hostID", "unique host id", "");

        try
        {
            op.parse(argc, argv);
        }
        catch (const std::invalid_argument& e)
        {
            cerr << "Exception: " << e.what() << std::endl;
            cout << "\n" << op << "\n";
            exit(EXIT_FAILURE);
        }

        if (versionSwitch->is_set())
        {
            cout << "snapclient v" << VERSION << "\n"
                 << "Copyright (C) 2014-2018 BadAix (snapcast@badaix.de).\n"
                 << "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n"
                 << "This is free software: you are free to change and redistribute it.\n"
                 << "There is NO WARRANTY, to the extent permitted by law.\n\n"
                 << "Written by Johannes M. Pohl.\n\n";
            exit(EXIT_SUCCESS);
        }

#ifdef HAS_ALSA
        if (listSwitch->is_set())
        {
            vector<PcmDevice> pcmDevices = AlsaPlayer::pcm_list();
            for (auto dev : pcmDevices)
            {
                cout << dev.idx << ": " << dev.name << "\n" << dev.description << "\n\n";
            }
            exit(EXIT_SUCCESS);
        }
#endif

        if (helpSwitch->is_set())
        {
            cout << op << "\n";
            exit(EXIT_SUCCESS);
        }

        if (groffSwitch->is_set())
        {
            GroffOptionPrinter option_printer(&op);
            cout << option_printer.print();
            exit(EXIT_SUCCESS);
        }

        if (instance <= 0)
            std::invalid_argument("instance id must be >= 1");


        // XXX: Only one metadata option must be set

        AixLog::Log::init<AixLog::SinkNative>("snapclient", AixLog::Severity::trace, AixLog::Type::special);
        if (debugOption->is_set())
        {
            AixLog::Log::instance().add_logsink<AixLog::SinkCout>(AixLog::Severity::trace, AixLog::Type::all, "%Y-%m-%d %H-%M-%S.#ms [#severity] (#tag_func)");
            if (!debugOption->value().empty())
                AixLog::Log::instance().add_logsink<AixLog::SinkFile>(AixLog::Severity::trace, AixLog::Type::all, debugOption->value(),
                                                                      "%Y-%m-%d %H-%M-%S.#ms [#severity] (#tag_func)");
        }
        else
        {
            AixLog::Log::instance().add_logsink<AixLog::SinkCout>(AixLog::Severity::info, AixLog::Type::all, "%Y-%m-%d %H-%M-%S [#severity]");
        }


        signal(SIGHUP, signal_handler);
        signal(SIGTERM, signal_handler);
        signal(SIGINT, signal_handler);

#ifdef HAS_DAEMON
        std::unique_ptr<Daemon> daemon;
        if (daemonOption->is_set())
        {
            string pidFile = "/var/run/snapclient/pid";
            if (instance != 1)
                pidFile += "." + cpt::to_string(instance);
            string user = "";
            string group = "";

            if (userValue->is_set())
            {
                if (userValue->value().empty())
                    std::invalid_argument("user must not be empty");

                vector<string> user_group = utils::string::split(userValue->value(), ':');
                user = user_group[0];
                if (user_group.size() > 1)
                    group = user_group[1];
            }
            daemon.reset(new Daemon(user, group, pidFile));
            daemon->daemonize();
            if (processPriority < -20)
                processPriority = -20;
            else if (processPriority > 19)
                processPriority = 19;
            if (processPriority != 0)
                setpriority(PRIO_PROCESS, 0, processPriority);
            SLOG(NOTICE) << "daemon started" << std::endl;
        }
#endif

        PcmDevice pcmDevice = getPcmDevice(soundcard);
#if defined(HAS_ALSA)
        if (pcmDevice.idx == -1)
        {
            cout << "soundcard \"" << soundcard << "\" not found\n";
            //			exit(EXIT_FAILURE);
        }
#endif

        if (host.empty())
        {
#if defined(HAS_AVAHI) || defined(HAS_BONJOUR)
            BrowseZeroConf browser;
            mDNSResult avahiResult;
            while (!g_terminated)
            {
                try
                {
                    if (browser.browse("_snapcast._tcp", avahiResult, 5000))
                    {
                        host = avahiResult.ip;
                        port = avahiResult.port;
                        if (avahiResult.ip_version == IPVersion::IPv6)
                            host += "%" + cpt::to_string(avahiResult.iface_idx);
                        LOG(INFO) << "Found server " << host << ":" << port << "\n";
                        break;
                    }
                }
                catch (const std::exception& e)
                {
                    SLOG(ERROR) << "Exception: " << e.what() << std::endl;
                }
                chronos::sleep(500);
            }
#endif
        }

        // Setup metadata handling
        std::shared_ptr<MetadataAdapter> meta;
        meta.reset(new MetadataAdapter);
        if (metaStderr)
            meta.reset(new MetaStderrAdapter);

        std::unique_ptr<Controller> controller(new Controller(hostIdValue->value(), instance, meta));
        if (!g_terminated)
        {
            LOG(INFO) << "Latency: " << latency << "\n";
            controller->start(pcmDevice, host, port, latency);
            while (!g_terminated)
                chronos::sleep(100);
            controller->stop();
        }
    }
    catch (const std::exception& e)
    {
        SLOG(ERROR) << "Exception: " << e.what() << std::endl;
        exitcode = EXIT_FAILURE;
    }

    SLOG(NOTICE) << "daemon terminated." << endl;
    exit(exitcode);
}
